### DISCUSSION TITLE
Unload Controlnet model from buffer?

### DISCUSSION THREAD

**Discussion by DragonHawkAlpha** [Mikubill/sd-webui-controlnet#1820] [General]
This code disables controlnet once used.  But I don't see a controlling API endpoint to disable.  Anyone know how to disable once used via txt2img API in Automatic1111 without server restart?

Code from Wiki does not have an endpoint in the API to run it.  I do not understand how this would be called either, where do you find StableDiffusionProcessing object?  Can someone add a disable endpoint?  Leaving out of txt2img request is not enough because it does not unload from buffer as far as I can workout.

import importlib
external_code = importlib.import_module('extensions.sd-webui-controlnet.scripts.external_code', 'external_code')

def disable_controlnet(p: StableDiffusionProcessing):
    external_code.update_cn_script_in_processing(p, [])

**Comment by ljleb** [comment]
This is code that should be put in a webui extension.

There is no endpoint to disable controlnet at the moment, but as long as you don't ask for controlnet to run through the API, it shouldn't do any processing. Just remove `"alwayson_scripts": {"controlnet": {"args": [...]}}` from the request and it should not use controlnet.

Does your use case require you to permanently disable controlnet for some reason?

**Reply by DragonHawkAlpha** [reply]
The "awayson_scripts" usage works just fine but once the script is removed Automatic1111 gets upset and throughs errors.  Once the script is in it has to be used or the server has to be restarted.  I think the ControlNet model is being buffered or something.  So I can use Automatic1111 API without ControlNet but after the first use of ControlNet, I then have to use ControlNet thereafter or restart the server.  I could try setting the weight to zero as an interim solution I guess.

**Reply by SethRobinson** [reply]
Yes, seeing this behavior too with my [AI Tools front end](https://github.com/SethRobinson/aitools_client) for auto1111.

After using ControlNet once (via api),  can no longer do a normal Text to Image generation at all (via api) due to it somehow still trying to use ControlNet despite not sending any controlnet parms via "alwayson_scripts", and gives an error that breaks all future renders.  (unless those renders are setup to use controlnet, of course)

  ```
 File "/home/glados/aitools_server/extensions/sd-webui-controlnet/scripts/batch_hijack.py", line 42, in processing_process_images_hijack
        return getattr(processing, '__controlnet_original_process_images_inner')(p, *args, **kwargs)
      File "/home/glados/aitools_server/modules/processing.py", line 743, in process_images_inner
        devices.test_for_nans(x, "vae")
      File "/home/glados/aitools_server/modules/devices.py", line 158, in test_for_nans
        raise NansException(message)
    modules.devices.NansException: A tensor with all NaNs was produced in VAE. This could be because there's not enough precision to represent the picture. Try adding --no-half-vae commandline argument to fix this. Use --disable-nan-check commandline argument to disable this check.
```

If the auto1111 server is reloaded *or* a new SD model is loaded (and not internally cached probably), it fixed it.

It seems to me there is probably a work around if we send controlnet API info and instruct it to turn off all controlnet units but I'm not sure how to do that, sending "none" or blank strings didn't work for me.

**Reply by ljleb** [reply]
I see, this is pretty odd. Passing the appropriate number of units with `"enabled": false` in each unit should be a workaround for now.

**Reply by DragonHawkAlpha** [reply]
I don't really understand the "enabled": false and where that would need to go?

**Reply by SethRobinson** [reply]
After updating to latest versions of auto1111 and ControlNet today, issue seems fixed completely, thanks!

**Reply by DragonHawkAlpha** [reply]
@ljleb I came back to this issue and your suggestion worked great! Thanks.
